### PR TITLE
Implemented MMR search for Redis

### PR DESCRIPTION
**Pull request by htahboub**
Description: Implemented MMR search for Redis. Pretty straightforward, just using the already implemented MMR method on similarity search–fetched docs.
Issue: #10059
Dependencies: None
Twitter handle: @hamza_tahboub

### CODE REVIEW

**Comment by baskaryan** [comment]
cc @Spartee @tylerhutcherson

**Comment by Spartee** [comment]
Will review tomorrow.

**Comment by HTahboub** [comment]
@Spartee had a chance to take a look at this?

**Comment by Spartee** [comment]
It's next up after these slides :)

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (nfcampos) [inline libs/langchain/langchain/vectorstores/redis/base.py:970]: Not possible to fetch the embeddings from Redis? Embeddings for docs have by definition been computed before

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (nfcampos) [inline libs/langchain/langchain/vectorstores/redis/base.py:974]: This is embedding the query twice, once inside `similarity_search` and a 2nd time in this line, we should do it only once

**Comment by HTahboub** [comment]
@nfcampos thanks, made a few changes to fix all of that, including implementing a `similarity_search_by_vector` function as has been done for some other vector scores, and I double-checked sure that everything still works. Let me know if everything looks all right now!

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (Spartee) [inline libs/langchain/langchain/vectorstores/redis/base.py:1057]: why the parameter name change here?

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (Spartee) [inline libs/langchain/langchain/vectorstores/redis/base.py:1023]: assign and return

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (Spartee) [inline libs/langchain/langchain/vectorstores/redis/base.py:884]: I think ``vector`` makes more sense unless it's defined in the parent class

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (Spartee) [inline libs/langchain/langchain/vectorstores/redis/base.py:1012]: there is a utility method for ``array_to_buffer``. Should we have a ``buffer_to_array``?

**Review by Spartee — COMMENTED** [review]
High level review, but want integration tests before I approve. Some comments

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (HTahboub) [reply libs/langchain/langchain/vectorstores/redis/base.py:1057]: I think it would be a more appropriate name given the change from a string query to just its embedding (a list).

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (HTahboub) [reply libs/langchain/langchain/vectorstores/redis/base.py:884]: Actually called `embedding` in the parent. I'll fix that.

**Comment by HTahboub** [comment]
@Spartee added some integration tests. Let me know if that looks okay.

**Comment by HTahboub** [comment]
Hey @Spartee, had a chance to take a look at the changes and tests yet?

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (Spartee) [inline libs/langchain/langchain/vectorstores/redis/base.py:991]: let's put something like "maximum distnace from the query vector "

**Review by Spartee — APPROVED** [review]
Looks good to me. one comment on docstrings. Tests pass when pulled locally. Nice work.

@baskaryan Good to go from my side if the CI tests pass here.

File: libs/langchain/langchain/vectorstores/redis/base.py
Comment (HTahboub) [reply libs/langchain/langchain/vectorstores/redis/base.py:991]: Sure, I'll change that for the other functions that have a distance threshold too

**Pull request merged**
